docs: Create documentation for update strategy (#224)
istio-ecosystem/sail-operator: docs/README.md

**Note:** The `IstioOperator` / `istioctl` example is separate from the Sail-operator. Setting `spec.components` or `spec.values.gateways` on your Sail-operator `Istio` resource **will not work**.

## Update Strategy

The sail-operator supports two update strategies to update the version of the Istio control plane: `InPlace` and `RevisionBased`. The default strategy is `InPlace`.

### InPlace

> When the `InPlace` strategy is used, the existing Istio control plane is updated in-place. The workloads therefore don't need to be moved from one control plane instance to another.

Reviewer: ```suggestion
When the `InPlace` strategy is used, the existing Istio control plane is replaced with a new version. The workload sidecars immediately connect to the new control plane. The workloads therefore don't need to be moved from one control plane instance to another.
```

#### Example using the InPlace strategy

Prerequisites:
* sail-operator needs to be installed.

> * `istioctl` is installed.

Reviewer: Also, why "sail-operator" instead of "Sail Operator"?

> * `istioctl` is installed.

Author: In all the document we are naming the operator as `Sail-operator`

> * `istioctl` is installed.

Reviewer: Why are we doing this? We're not calling it "Sail-operator", but "sail-operator". But both forms look off. They look like a binary name or deployment/pod name. But that's an implementation detail. We're talking about the operator, not binary/deployment/pod. And the operator's name is "Sail Operator". That's also how it's listed in the OperatorHub catalog.

> * Deploy `IstioCni` resource.

Reviewer: IstioCNI is not a prerequisite. Since these aren't OpenShift-specific docs, we can skip IstioCNI completely.

* `istioctl` is installed.

Steps:
1. Create the `istio-system` namespace.

    ```bash
    kubectl create namespace istio-system
    ```

2. Create the `Istio` resource

    ```bash
    cat <<EOF | kubectl apply -f-
    apiVersion: operator.istio.io/v1alpha1
    kind: Istio
    metadata:
      name: default
    spec:
      namespace: istio-system
      updateStrategy:
        type: InPlace

>         inactiveRevisionDeletionGracePeriodSeconds: 30

Reviewer: I suggest we skip this line to make the example simpler.

      version: v1.21.0
    EOF
    ```

3. Confirm the installation and version of the control plane.

    ```bash
    $ kubectl get istio -n istio-system
    NAME      READY   STATUS    IN USE   VERSION   AGE
    default   True    Healthy   True     v1.21.0   2m
    ```

4. Create namespace `bookinfo` and deploy bookinfo application

    ```bash
    kubectl create namespace bookinfo
    kubectl label namespace bookinfo istio-injection=enabled
    kubectl apply -n bookinfo -f https://raw.githubusercontent.com/istio/istio/release-1.22/samples/bookinfo/platform/kube/bookinfo.yaml
    ```

> Note: if the `Istio` resource name is other than default, you need to label the namespace with the revision name.

Reviewer: ```suggestion
Note: if the `Istio` resource name is other than `default`, you need to set the `istio.io/rev` label to the name of the 
`Istio` resource instead of adding the `istio-injection=enabled` label.
```

5. Perform the update of the control plane by changing the version in the Istio resource

    ```bash
    kubectl patch istio default -n istio-system --type='merge' -p '{"spec":{"version":"v1.21.2"}}'
    ```

> 6. Confirm the istio resource version was updated

Reviewer: ```suggestion
6. Confirm the `Istio` resource version was updated
```

    ```bash
    $ kubectl get istio -n istio-system
    NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   VERSION   AGE
    default   1           1       1        Healthy           v1.21.2   12m
    ```

> 7. Delete `bookinfo` pod to trigger sidecar injection with the new version

Reviewer: ```suggestion
7. Delete `bookinfo` pods to trigger sidecar injection with the new version
```

    kubectl rollout restart deployment -n bookinfo
    ```

8. Confirm that the new version is used in the sidecar

    ```bash
    istioctl proxy-status 
    ```
The column `VERSION` should match the new control plane version.

### RevisionBased

> When the `RevisionBased` strategy is used, a new Istio control plane instance is created for every change to the `Istio.spec.version` field. The old control plane remains in place until all workloads have been moved to the new control plane instance, then, the old control plane is deleted.

Reviewer: I'd rephrase the second sentence to make it clear that the user is the one that has to move the workloads to the new control plane instance. Currently, some readers could misunderstand the sentence to mean that all this happens automatically.

### RevisionBased
When the `RevisionBased` strategy is used, a new Istio control plane instance is created for every change to the `Istio.spec.version` field. The old control plane remains in place until all workloads have been moved to the new control plane instance, then, the old control plane is deleted.

#### Example using the RevisionBased strategy

Prerequisites:
* sail-operator needs to be installed.
* Deploy `IstioCni` resource.
* `istioctl` is installed.

Steps:

1. Create the `istio-system` namespace.

    ```bash
    kubectl create namespace istio-system
    ```

2. Create the `Istio` resource

    ```bash
    cat <<EOF | kubectl apply -f-
    apiVersion: operator.istio.io/v1alpha1
    kind: Istio
    metadata:
      name: default
    spec:
      namespace: istio-system
      updateStrategy:

>         type: InPlace

Reviewer 2: ```suggestion
        type: RevisionBased
```

>         inactiveRevisionDeletionGracePeriodSeconds: 30

Reviewer: Skip this line.

Reviewer: Actually, no. You mention this grace period later, so let's keep it here. But maybe still remove it from the InPlace example.

      version: v1.21.0
    EOF
    ```

3. Confirm the control plane is installed and is using the desired version.

    ```bash
    $ kubectl get istio -n istio-system
    NAME      READY   STATUS    IN USE   VERSION   AGE
    default   True    Healthy   True     v1.21.0   2m
    ```

> 4. Get the `istiorevision` name

Reviewer: ```suggestion
4. Get the `IstioRevision` name
```

Please also check capitalization everywhere else.

    ```bash
    $ kubectl get istiorevision -n istio-system
    NAME              READY   STATUS    IN USE   VERSION   AGE
    default-v1-21-0   True    Healthy   False    v1.21.0   114s

>     ```

Reviewer: Maybe add note saying that the revision name is composed of the `Istio` resource name and the version.

    ```bash
    kubectl create namespace bookinfo
    kubectl label namespace bookinfo istio.io/rev=default-v1-21-0
    ```

6. Deploy bookinfo application

    ```bash
    kubectl apply -n bookinfo -f https://raw.githubusercontent.com/istio/istio/release-1.22/samples/bookinfo/platform/kube/bookinfo.yaml
    ```

7. Confirm proxy sidecar injection version match the control plane version

    ```bash
    istioctl proxy-status 
    ```
The column `VERSION` should match the control plane version.

8. Update the control plane to a new version

    ```bash
    kubectl patch istio default -n istio-system --type='merge' -p '{"spec":{"version":"v1.21.2"}}'
    ```

9. Verify the `Istio` and `IstioRevision` resources. There will be a new revision created with the new version.

    ```console

>     $ kubectl get istio -n istio-system
>       NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   VERSION   AGE
>       default   2           2       1        Healthy           v1.21.2   23m
>     $ kubectl get istiorevision -n istio-system
>       NAME              READY   STATUS    IN USE   VERSION   AGE
>       default-v1-21-0   True    Healthy   True     v1.21.0   27m
>       default-v1-21-2   True    Healthy   False    v1.21.2   4m45s

Reviewer: Indentation:

```suggestion
    $ kubectl get istio -n istio-system
    NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   VERSION   AGE
    default   2           2       1        Healthy           v1.21.2   23m

    $ kubectl get istiorevision -n istio-system
    NAME              READY   STATUS    IN USE   VERSION   AGE
    default-v1-21-0   True    Healthy   True     v1.21.0   27m
    default-v1-21-2   True    Healthy   False    v1.21.2   4m45s
```

    ```

10. Confirm there are two control plane pods running, one for each revision

    ```console

>     $ kubectl get pods -n istio-system
>       NAME                                      READY   STATUS    RESTARTS   AGE
>       istiod-default-v1-21-0-69d6df7f9c-grm24   1/1     Running   0          28m
>       istiod-default-v1-21-2-7c4f4674c5-4g7n7   1/1     Running   0          6m9s

Reviewer: ```suggestion
    $ kubectl get pods -n istio-system
    NAME                                      READY   STATUS    RESTARTS   AGE
    istiod-default-v1-21-0-69d6df7f9c-grm24   1/1     Running   0          28m
    istiod-default-v1-21-2-7c4f4674c5-4g7n7   1/1     Running   0          6m9s
```

      istiod-default-v1-21-0-69d6df7f9c-grm24   1/1     Running   0          28m
      istiod-default-v1-21-2-7c4f4674c5-4g7n7   1/1     Running   0          6m9s
    ```

11. Confirm the proxy sidecar version remains the same:

    ```bash
    istioctl proxy-status 
    ```
The column `VERSION` should still match the old control plane version.

12. Change the label of the `bookinfo` namespace to use the new revision

    ```bash
    kubectl label namespace bookinfo istio.io/rev=default-v1-21-2 --overwrite
    ```

> The injected version by the control plane will not change immediately. Only will be changed after the pods are restarted, this will trigger the sidecar injection with the new version.

Reviewer: ```suggestion
The injected version by the control plane will not change immediately. It will only be changed after the pods are restarted, this will trigger the sidecar injection with the new version.
```

> The injected version by the control plane will not change immediately. It will only be changed after the pods are restarted, this will trigger the sidecar injection with the new version.

Reviewer: ```suggestion
The existing workload sidecars will continue to run and will remain connected to the old control plane instance. The sidecars will not be replaced with a new version until the pods are deleted and recreated.
```

13. Delete all the pods in the `bookinfo` namespace

    ```bash
    kubectl rollout restart deployment -n bookinfo
    ```

14. Confirm the new version is used in the sidecars

    ```bash
    istioctl proxy-status 
    ```
The column `VERSION` should match the updated control plane version.

> 15. Confirm the old control plane deletion
> The old control plane will be deleted after the grace period specified in the `Istio` resource `Istio.spec.updateStrategy.inactiveRevisionDeletionGracePeriodSeconds`.

Reviewer: ```suggestion
15. Confirm the old control plane deletion

The old control plane will be deleted after the grace period specified in the `Istio` resource `Istio.spec.updateStrategy.inactiveRevisionDeletionGracePeriodSeconds`.
```

> The old control plane will be deleted after the grace period specified in the `Istio` resource `Istio.spec.updateStrategy.inactiveRevisionDeletionGracePeriodSeconds`.

Reviewer 2: nit: Since we are kind of repeating similar message even in Line 327, we can probably club number 15 and 16 together and show both the outputs to avoid the duplication.

    ```bash
    $ kubectl get pods -n istio-system
    NAME                                      READY   STATUS    RESTARTS   AGE
    istiod-default-v1-21-2-7c4f4674c5-4g7n7   1/1     Running   0          94m
    ```

> 16. Confirm the old revision deletion
> The old revision will be deleted after the grace period specified in the `Istio` resource `Istio.spec.updateStrategy.inactiveRevisionDeletionGracePeriodSeconds`.

Reviewer: ```suggestion
16. Confirm the old revision deletion

The old revision will be deleted after the grace period specified in the `Istio` resource `Istio.spec.updateStrategy.inactiveRevisionDeletionGracePeriodSeconds`.
```

> The old revision will be deleted after the grace period specified in the `Istio` resource `Istio.spec.updateStrategy.inactiveRevisionDeletionGracePeriodSeconds`.

Reviewer: The indentation doesn't look alright here and in some of the previous bullets.

Reviewer: Also, the code block seems to be broken here and in step 15.

**Note:** The `IstioOperator` / `istioctl` example is separate from the Sail-operator. Setting `spec.components` or `spec.values.gateways` on your Sail-operator `Istio` resource **will not work**.

## Update Strategy

> The sail-operator supports two update strategies to update the istio version of the control plane: `InPlace` and `RevisionBased`. The default strategy is `InPlace`.

Reviewer 3: maybe: 'to update a version of the Istio control plane'?

### InPlace
When the `InPlace` strategy is used, the existing Istio control plane is updated in-place. The workloads therefore don't need to be moved from one control plane instance to another.

#### Example using the InPlace strategy

Prerequisites:
* sail-operator needs to be installed.

Steps:
1. Create the `istio-system` and the `istio-cni` namespaces.

    ```bash
    kubectl create namespace istio-system
    kubectl create namespace istio-cni
    ```

> 2. Create the `Istio-cni` resource

Reviewer: ```suggestion
2. Create the `IstioCni` resource
```

Reviewer: But, I think we should omit `IstioCNI` here altogether. IstioCNI is not needed and mentioning it here only makes the important (upgrade) stuff less visible.

Author: I deleted the section, we are going to skip mentioning the istioCNI resource

  
      ```bash
      cat <<EOF | kubectl apply -f -
      apiVersion: operator.istio.io/v1alpha1
      kind: IstioCNI
      metadata:
        name: default
      spec:
        version: v1.21.0
        namespace: istio-cni
      EOF
      ```

3. Create the `Istio` resource

    ```bash
    cat <<EOF | kubectl apply -f-
    apiVersion: operator.istio.io/v1alpha1
    kind: Istio
    metadata:
      name: default
    spec:
      namespace: istio-system
      updateStrategy:
        type: InPlace
        inactiveRevisionDeletionGracePeriodSeconds: 30
      version: v1.21.0
    EOF
    ```

4. Confirm the installation and version of the control plane.

    ```bash
    $ kubectl get istio -n istio-system
    NAME      READY   STATUS    IN USE   VERSION   AGE
    default   True    Healthy   True     v1.21.0   2m
    ```

5. Create namespace `bookinfo` and deploy bookinfo application

    ```bash
    kubectl create namespace bookinfo

>     kubectl label namespace bookinfo istio.io/rev=default

Author: For `InPlace` strategy we can use both labels or `istio-io/rev` or `istio-injection`. For `InPlace` make sense to use only `istio-injection=enabled`

>     kubectl label namespace bookinfo istio.io/rev=default

Reviewer: Be careful. Even if you use `InPlace`, you have to use `istio.io/rev` if the name of your `Istio` CR is anything other than `default`.

    kubectl apply -n bookinfo -f https://raw.githubusercontent.com/istio/istio/release-1.22/samples/bookinfo/platform/kube/bookinfo.yaml
    ```

6. Perform the update of the control plane by changing the version in the Istio resource

    ```bash

>     cat <<EOF | kubectl apply -f-

Reviewer: +1

Use patch to update only the `spec.version` field.

    apiVersion: operator.istio.io/v1alpha1
    kind: Istio
    metadata:
      name: default
    spec:
      namespace: istio-system
      updateStrategy:
        type: InPlace
        inactiveRevisionDeletionGracePeriodSeconds: 30
      version: v1.21.2
    EOF
    ```

7. Confirm the istio resource version was updated

    ```bash

>     kubectl get istio -n istio-system
>     NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   VERSION   AGE
>     default   1           1       1        Healthy           v1.21.2   12m

Reviewer: ```suggestion
    $ kubectl get istio -n istio-system
    NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   VERSION   AGE
    default   1           1       1        Healthy           v1.21.2   12m
```

Or remove the `$` from the other code blocks above.

    ```

8. Delete `bookinfo` pod to trigger sidecar injection with the new version

    ```bash
    kubectl delete pods --all -n bookinfo
    ```

9. Confirm that the new version is used in the sidecar

    ```bash

>     kubectl get pod -n bookinfo -l app=details -o yaml | yq '.items[].spec.containers[].image'
>     docker.io/istio/examples-bookinfo-details-v1:1.19.1
>     docker.io/istio/proxyv2:1.21.2

Reviewer: ```suggestion
    $ kubectl get pod -n bookinfo -l app=details -o yaml | yq '.items[].spec.containers[].image'
    docker.io/istio/examples-bookinfo-details-v1:1.19.1
    docker.io/istio/proxyv2:1.21.2
```

This makes it easy to differentiate between the command and its output.

>     docker.io/istio/proxyv2:1.21.2

Reviewer: Or just check the annotation (I think it's `istio.io/rev` or something similar).

>     docker.io/istio/proxyv2:1.21.2

Author: Using istioctl is the best option because there is no label with the istio version used

>     docker.io/istio/proxyv2:1.21.2

Reviewer: But do you really need to see the version? Or do you actually want to see the *revision* that the pod is attached to?

```
kubectl get pod -n bookinfo -l app=details -oyaml | yq '.items[].metadata.annotations["istio.io/rev"]'
```

    ```

### RevisionBased

> When the `RevisionBased` strategy is used, a new Istio control plane instance is created for every change to the Istio.spec.version field. The old control plane remains in place until all workloads have been moved to the new control plane instance, then, the old control plane is deleted.

Reviewer 3: Is the old control plane removed automatically?

Author: Yes, it will be removed automatically when all the pods are moved to the new one

        inactiveRevisionDeletionGracePeriodSeconds: 30
      version: v1.21.0
    EOF
    ```

4. Confirm the control plane is installed and is using the desired version.

    ```bash
    $ kubectl get istio -n istio-system
    NAME      READY   STATUS    IN USE   VERSION   AGE
    default   True    Healthy   True     v1.21.0   2m
    ```

5. Get the `istiorevision` name

    ```bash
    $ kubectl get istiorevision -n istio-system
    NAME              READY   STATUS    IN USE   VERSION   AGE
    default-v1-21-0   True    Healthy   False    v1.21.0   114s
    ```

6. Create `bookinfo` namespace and label it with the revision name

    ```bash
    kubectl create namespace bookinfo
    kubectl label namespace bookinfo istio.io/rev=default-v1-21-0
    ```

7. Deploy bookinfo application

    ```bash
    kubectl apply -n bookinfo -f https://raw.githubusercontent.com/istio/istio/release-1.22/samples/bookinfo/platform/kube/bookinfo.yaml
    ```

8. Confirm proxy sidecar injection version match the control plane version

    ```bash
    kubectl get pod -n bookinfo -l app=details -o yaml | yq '.items[].spec.containers[].image'
    docker.io/istio/examples-bookinfo-details-v1:1.19.1

>     docker.io/istio/proxyv2:1.21.0

Reviewer 3: The same as above for in-place

    ```

9. Update the control plane to a new version

    ```bash
    cat <<EOF | kubectl apply -f-
    apiVersion: operator.istio.io/v1alpha1
    kind: Istio
    metadata:
      name: default
    spec:
      namespace: istio-system
      updateStrategy:
        type: RevisionBased
        inactiveRevisionDeletionGracePeriodSeconds: 30
      version: v1.21.2
    EOF
    ```

10. Verify the `Istio` and `IstioRevision` resources. There will be a new revision created with the new version.

    ```bash
    $ kubectl get istio -n istio-system
      NAME      REVISIONS   READY   IN USE   ACTIVE REVISION   VERSION   AGE
      default   2           2       1        Healthy           v1.21.2   23m
    $ kubectl get istiorevision -n istio-system
      NAME              READY   STATUS    IN USE   VERSION   AGE
      default-v1-21-0   True    Healthy   True     v1.21.0   27m
      default-v1-21-2   True    Healthy   False    v1.21.2   4m45s
    ```

> 11. Confirm there is two control plane pods running, one for each revision

Reviewer 2: ```suggestion
11. Confirm there are two control plane pods running, one for each revision
```

Reviewer: @sridhargaddam We do have tests that check the grace period. Are you sure they're getting deleted immediately and not after 30s (the default grace period)?

Reviewer 2: @luksa I was following the steps mentioned in this PR and I'm fairly sure of seeing this behavior. 

However, I believe the reason is that when the initial Istio object was created at Line 248, it was given the name default. When it is updated at Line 302, we are merely updating the same Istio object. 

Shouldn't we use a different name for `Istio` objects when using revisions?

Reviewer: No, using the same object is correct. The operator takes care of creating two `IstioRevision` objects underneath.

Reviewer 2: There was some confusion on my end. I misread it as 30 mins and not seconds. We are good. Thanks guys!

    ```bash
    $ kubectl get pods -n istio-system
      NAME                                      READY   STATUS    RESTARTS   AGE
      istiod-default-v1-21-0-69d6df7f9c-grm24   1/1     Running   0          28m
      istiod-default-v1-21-2-7c4f4674c5-4g7n7   1/1     Running   0          6m9s
    ```

12. Confirm the proxy sidecar version remains the same:

    ```bash
    $ kubectl get pod -n bookinfo -l app=details -o yaml | yq '.items[].spec.containers[].image'
      docker.io/istio/examples-bookinfo-details-v1:1.19.1
      docker.io/istio/proxyv2:1.21.0
    ```

13. Change the label of the `bookinfo` namespace

    ```bash
    kubectl label namespace bookinfo istio.io/rev=default-v1-21-2 --overwrite
    ```

> The control plane used by the pods will be updated to the new version only after all workloads are moved to the new revision, and this will happen after the pods are restarted.

Author: Done

14. Delete all the pods in the `bookinfo` namespace

    ```bash
    kubectl delete pods --all -n bookinfo
    ```

15. Confirm the new version is used in the sidecars

    ```bash
    kubectl get pod -n bookinfo -l app=details -o yaml | yq '.items[].spec.containers[].image'
    docker.io/istio/examples-bookinfo-details-v1:1.19.1
    docker.io/istio/proxyv2:1.21.2
    ```

> 16. Confirm the old control plane deletion

Reviewer: Add note saying that the old revision (and control plane) are deleted after the grace period specified in `Istio.spec.updateStrategy.inactiveRevisionDeletionGracePeriodSeconds` runs out.

Reviewer 2: @luksa I remember seeing that after we update the label on bookinfo namespace to point to new revision and its restarted, the old Istio revision (and control plane) pods are immediately deleted. i.e., it does not wait for `inactiveRevisionDeletionGracePeriodSeconds` 

Do you think its a bug?
